src/main/scala/internal/Graph.scala

```diff
@@ -0,0 +1,425 @@
+    copy(dependencies = newDependencies)
+  }
+
+  /**
+   * Immutably update the operator of a node in this graph.
+   *
+   * @param node The id of the node to be updated
+   * @param op The new operator to assign to the node
+   * @return The new graph
+   */
+  def setOperator(node: NodeId, op: Operator): Graph = {
+    require(dependencies.contains(node), "Node being updated must exist")
+
+    val newOperators = operators.updated(node, op)
+    copy(operators = newOperators)
+  }
+
+  /**
+   * Immutably update the dependency of a sink in this graph.
+   *
+   * @param sink The id of the sink to be updated
+   * @param dep The new dependency to assign to the sink
+   * @return The new graph
+   */
+  def setSinkDependency(sink: SinkId, dep: NodeOrSourceId): Graph = {
+    require(sinkDependencies.contains(sink), "Sink being updated must exist")
+
+    require ({
+      val nodesAndSources: Set[NodeOrSourceId] = nodes ++ sources
+      nodesAndSources.contains(dep)
+    }, "Sink must have dependencies on an existing id")
+
+    val newSinkDependencies = sinkDependencies.updated(sink, dep)
+    copy(sinkDependencies = newSinkDependencies)
+  }
+
+  /**
+   * Immutably remove a sink from this graph.
+   *
+   * @param sink The id of the sink to remove
+   * @return The new graph
+   */
+  def removeSink(sink: SinkId): Graph = {
+    require(sinkDependencies.contains(sink), "Sink being removed must exist")
+
+    val newSinkDependencies = sinkDependencies - sink
+    copy(sinkDependencies = newSinkDependencies)
+  }
+
+  /**
+   * Immutably remove a source from this graph.
+   * Note: This may leave invalid dangling dependencies on the deleted source,
+   * that must be manually taken care of.
+   *
+   * @param source The id of the source to remove
+   * @return The new graph
+   */
+  def removeSource(source: SourceId): Graph = {
+    require(sources.contains(source), "Source being removed must exist")
+
+    val newSources = sources - source
+    copy(sources = newSources)
+  }
+
+  /**
+   * Immutably remove a node from this graph.
+   * Note: This may leave invalid dangling dependencies on the deleted node,
+   * that must be manually taken care of.
+   *
+   * @param node The id of the node to remove
+   * @return The new graph
+   */
+  def removeNode(node: NodeId): Graph = {
+    require(nodes.contains(node), "Node being removed must exist")
+
+    val newOperators = operators - node
+    val newDependencies = dependencies - node
+    copy(operators = newOperators, dependencies = newDependencies)
+  }
+
+  /**
+   * Immutably replace all dependencies on a given node or source with a new target.
+   *
+   * @param oldDep The id of the old node or source that is the existing dependency target
+   * @param newDep The id of the node or source intended to be the new target
+   * @return The new graph
+   */
+  def replaceDependency(oldDep: NodeOrSourceId, newDep: NodeOrSourceId): Graph = {
```

Comment:
`replaceDependency` actually doesn't affect id space tracking, because it isn't assigning new ids to existing nodes or sources. It's just moving dependencies around between existing nodes w/o modifying any ids.

Only `addGraph` and `connectGraph` add id space bookkeeping (and I've made sure to isolate the bookkeeping to sources and sinks, instead of having to do node id bookkeeping as well).